components/brave_perf_predictor/browser/bandwidth_savings_predictor.cc

```diff
@@ -0,0 +1,151 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_perf_predictor/browser/bandwidth_savings_predictor.h"
+
+#include "base/logging.h"
+#include "brave/components/brave_perf_predictor/browser/bandwidth_linreg.h"
+#include "brave/components/brave_perf_predictor/browser/third_party_extractor.h"
+#include "content/public/common/resource_type.h"
+#include "net/base/registry_controlled_domains/registry_controlled_domain.h"
+
+namespace brave_perf_predictor {
+
+BandwidthSavingsPredictor::BandwidthSavingsPredictor() {}
+
+BandwidthSavingsPredictor::~BandwidthSavingsPredictor() = default;
+
+void BandwidthSavingsPredictor::OnPageLoadTimingUpdated(
+    const page_load_metrics::mojom::PageLoadTiming& timing) {
+  // First meaningful paint
+  if (timing.paint_timing->first_meaningful_paint.has_value())
+    feature_map_["metrics.firstMeaningfulPaint"] =
+        timing.paint_timing->first_meaningful_paint.value().InMillisecondsF();
+
+  // DOM Content Loaded
+  if (timing.document_timing->dom_content_loaded_event_start.has_value())
+    feature_map_["metrics.observedDomContentLoaded"] =
+        timing.document_timing->dom_content_loaded_event_start.value()
+            .InMillisecondsF();
+
+  // First contentful paint
+  if (timing.paint_timing->first_contentful_paint.has_value())
+    feature_map_["metrics.observedFirstVisualChange"] =
+        timing.paint_timing->first_contentful_paint.value().InMillisecondsF();
+
+  // Load
+  if (timing.document_timing->load_event_start.has_value())
+    feature_map_["metrics.observedLoad"] =
+        timing.document_timing->load_event_start.value().InMillisecondsF();
+
+  // Interactive
+  if (timing.interactive_timing->interactive.has_value())
+    feature_map_["metrics.interactive"] =
+        timing.interactive_timing->interactive.value().InMillisecondsF();
+}
+
+void BandwidthSavingsPredictor::OnSubresourceBlocked(
+    const std::string& resource_url) {
+  feature_map_["adblockRequests"] += 1;
+
+  ThirdPartyExtractor* extractor = ThirdPartyExtractor::GetInstance();
+
+  auto entity_name = extractor->GetEntity(resource_url);
+  if (entity_name.has_value())
+    feature_map_["thirdParties." + entity_name.value() + ".blocked"] = 1;
+}
+
+void BandwidthSavingsPredictor::OnResourceLoadComplete(
+    const GURL& main_frame_url,
+    const content::mojom::ResourceLoadInfo& resource_load_info) {
+  // If the resource load info comes without a valid corresponding
+  // main frame URL, ignore it
```

Comment:
not sure, but probably we will skip data that come from service workers this way?

Reply:
possibly, but the prediction model was not including service workers either, only an approximation in any case. Would leave for next iteration